### PR TITLE
Use CRS in WMS GetFeatureInfo when info_format is in json Fixes #32326

### LINKED ISSUE #32326: QGIS Server GetFeatureInfo request: if INFO_FORMAT=application/json CRS is not respected
**Describe the bug**

QGIS Server handles `GetFeatureInfo` requests. The mandatory `INFO_FORMAT` parameter indicates what format to use when returning the feature information.

For the same request:
* if `INFO_FORMAT=application/json` the result is in EPSG:4326 coordinates
* if `INFO_FORMAT=text/xml` the result is according to the CRS parameter

**How to Reproduce**

```
curl "http://qgis.demo/cgi-bin/qgis_mapserv.fcgi?MAP=postgresql/geotuga/public/domingo&SERVICE=WMS&VERSION=1.3.0&REQUEST=GetFeatureInfo&FORMAT=image%2Fpng&TRANSPARENT=true&QUERY_LAYERS=npolicia&tiled=true&STYLES=default&LAYERS=npolicia&CRS=EPSG%3A3763&INFO_FORMAT=application%2Fjson&FEATURE_COUNT=50&WITH_GEOMETRY=True&I=222&J=127&WIDTH=256&HEIGHT=256&BBOX=-40866.001953125%2C120021.9140625%2C-40830.5859375%2C120057.330078125"
```

```json
{
  "features": [
    {
      "bbox": [
        -8.616624,
        40.748309,
        -8.616624,
        40.748309
      ],
      "geometry": {
        "coordinates": [
          [
            -8.616624,
            40.748309
          ]
        ],
        "type": "MultiPoint"
      },
      "id": "npolicia.10494",
      "properties": {
        "artigo": "010807-U-00753 (171689321)",
        "cil2": "0",
        "codine": "010807",
        "codpostal": "3860-645",
        "colocacao": "Sim",
        "colocado_2": null,
        "comentario": "PROCESSO",
        "descricao": null,
        "dicofre": "010808",
        "id": 10494,
        "ligacacao": null,
        "nome_da_ru": "Rua de Stº António",
        "num_via": 75,
        "numero": "24",
        "observacoes": null,
        "registo": null,
        "t_local": "8684464",
        "tipo": "DOMESTICO"
      },
      "type": "Feature"
    }
  ],
  "type": "FeatureCollection"
}
```

```
curl "http://qgis.demo/cgi-bin/qgis_mapserv.fcgi?MAP=postgresql/geotga/public/domingo&SERVICE=WMS&VERSION=1.3.0&REQUEST=GetFeatureInfo&FORMAT=image%2Fpng&TRANSPARENT=true&QUERY_LAYERS=npolicia&tiled=true&STYLES=default&LAYERS=npolicia&CRS=EPSG%3A3763&INFO_FORMAT=text%2Fxml&FEATURE_COUNT=50&WITH_GEOMETRY=True&I=222&J=127&WIDTH=256&HEIGHT=256&BBOX=-40866.001953125%2C120021.9140625%2C-40830.5859375%2C120057.330078125&_dc=1571591331252"
```

```xml
<GetFeatureInfoResponse>
 <Layer name="npolicia">
  <Feature id="10494">
   <Attribute name="id" value="10494"/>
   <Attribute name="numero" value="24"/>
   <Attribute name="dicofre" value="010808"/>
   <Attribute name="descricao" value=""/>
   <Attribute name="registo" value=""/>
   <Attribute name="nome_da_ru" value="Rua de Stº António"/>
   <Attribute name="colocacao" value="Sim"/>
   <Attribute name="comentario" value="PROCESSO"/>
   <Attribute name="codpostal" value="3860-645"/>
   <Attribute name="tipo" value="DOMESTICO"/>
   <Attribute name="num_via" value="75"/>
   <Attribute name="cil2" value="0"/>
   <Attribute name="t_local" value="8684464"/>
   <Attribute name="artigo" value="010807-U-00753 (171689321)"/>
   <Attribute name="codine" value="010807"/>
   <Attribute name="colocado_2" value=""/>
   <Attribute name="ligacacao" value=""/>
   <Attribute name="observacoes" value=""/>
   <BoundingBox maxy="120039.83261115" CRS="EPSG:3763" miny="120039.83261115" minx="-40835.12320317" maxx="-40835.12320317"/>
   <Attribute name="geometry" value="MultiPoint ((-40835.12320317 120039.83261115))" type="derived"/>
  </Feature>
 </Layer>
</GetFeatureInfoResponse>
```

**QGIS and OS versions**

QGIS Master, Ubuntu 18.04

**Additional context**

Using `INFO_FORMAT=application/json` we are just encoding the result in JSON, not in GEOJson (there is a mime type for geojson, which is `application/geo+json`).

Geoserver returns the result is according to the CRS parameter, if `INFO_FORMAT=application/json`.

### PR DESCRIPTION
**Pull request by pblottiere**
## Description

Fixes https://github.com/qgis/QGIS/issues/32326

### CODE REVIEW

File: src/core/qgsjsonutils.h [hunk 1]
```diff
@@ -54,6 +54,8 @@ class CORE_EXPORT QgsJsonExporter
      */
     QgsJsonExporter( QgsVectorLayer *vectorLayer = nullptr, int precision = 6 );
 
+    void setDestinationCrs( const QgsCoordinateReferenceSystem &crs );;
```
Comment (elpaso) [inline src/core/qgsjsonutils.h]:
```suggestion
    void setDestinationCrs( const QgsCoordinateReferenceSystem &crs );
```

Comment (pblottiere) [reply src/core/qgsjsonutils.h]: oops, thanks

**Comment by jgrocha** [comment]
Good work @pblottiere 

### The good news

Tested locally and it works perfectly with `INFO_FORMAT=application/json` and `CRS=EPSG:3763`. The project was in `CRS=EPSG:3763`.

### Not so good news

It is odd, but the same project, if I request `INFO_FORMAT=application/json` and `CRS=EPSG:3857` the result goes to the north pole. If I request `INFO_FORMAT=text/xml` the result is fine.

### To reproduce

Using the previous project stored in the `gfi.gpkg` (and running with your PR):

```bash
export QGIS_PROJECT_FILE="geopackage:/home/qgis/projects/gfi.gpkg?projectName=testgfi"
export QUERY_STRING="SERVICE=WMS&VERSION=1.3.0&REQUEST=GetFeatureInfo&FORMAT=image/png&TRANSPARENT=true&tiled=false&STYLES=default&LAYERS=freguesias&CRS=EPSG:3857&WIDTH=512&HEIGHT=512&BBOX=-953087.8,4973324.9,-952748.9,4973612.8&QUERY_LAYERS=freguesias&INFO_FORMAT=application/json&FEATURE_COUNT=50&WITH_GEOMETRY=True&I=222&J=127"
./build-QGIS-Desktop-Debug/output/bin/qgis_mapserv.fcgi
```

The request uses EPSG:3857 coordinates, around -953 087, 4 973 324.

The coordinates in JSON are around -7 194 793, 15 428 040. 

```json
{
  "features": [
    {
      "bbox": [
        -7202100.975262,
        15370117.593422,
        -7179015.488842,
        15428040.830143001
      ],
      "geometry": {
        "coordinates": [
          [
            [
              [
                -7194793.711287,
                15428040.830143001
              ],
              [
                -7194757.735689,
                15428031.546399
              ],
```

I've attached the resulting file [x.json](https://github.com/qgis/QGIS/files/3775994/x.zip). If you open it with QGIS (and assign the EPSG:3857), it goes to the north pole.

Unless it is a local issue (or my mistake), when the output is json, the conversion might not work as expected.

**Comment by pblottiere** [comment]
Hi @jgrocha,

> It is odd, but the same project, if I request INFO_FORMAT=application/json and CRS=EPSG:3857 the result goes to the north pole. If I request INFO_FORMAT=text/xml the result is fine.

Ah yes, I forgot the second part of your issue, sorry for that.

> The request uses EPSG:3857 coordinates, around -953 087, 4 973 324.
> The coordinates in JSON are around -7 194 793, 15 428 040.

I think it's fixed now. In both cases we have the same coordinates (excepted for the precision which is slightly different).

#### INFO_FORMAT=text/xml and CRS=EPSG:3857

```
value="MultiPolygon (((-949439.66200299 4977488.4941173, -949437.44500259 4977482.21731793, -949437.14610265 4977465.63576632, ....
```

#### INFO_FORMAT=application/json and CRS=EPSG:3857

```
"geometry":{"coordinates":[[[[-949439.662003,4977488.494117],[-949437.445003,4977482.217318],[-949437.146103,4977465.635766],
```

Can you take a look locally with your own project to make sure that everything is OK please?

**Comment by jgrocha** [comment]
Hi @pblottiere,

Thanks for keeping working on this PR. The example with EPSG:3857 is working now. Nice work!

But... but we still have issues with CRS transformations. I've made a request with EPSG:4326 coordinates and the result is an empty image. The GetMap fails (and GetFeatureInfo as well).

Please test it, because it might be my fault.

For the very same project:

```bash
export QUERY_STRING="SERVICE=WMS&VERSION=1.3.0&REQUEST=GetMap&FORMAT=image/png&TRANSPARENT=true&tiled=false&STYLES=default&LAYERS=freguesias&CRS=EPSG:4326&WIDTH=512&HEIGHT=512&BBOX=-8.61,40.70,-8.53,40.77"
./build-QGIS-Desktop-Debug/output/bin/qgis_mapserv.fcgi > image.png
```

The image is empty!

```bash
export QUERY_STRING="SERVICE=WMS&VERSION=1.3.0&REQUEST=GetFeatureInfo&FORMAT=image/png&TRANSPARENT=true&tiled=false&STYLES=default&LAYERS=freguesias&CRS=EPSG:4326&WIDTH=512&HEIGHT=512&BBOX=-8.61,40.70,-8.53,40.77&QUERY_LAYERS=freguesias&INFO_FORMAT=application/json&FEATURE_COUNT=50&WITH_GEOMETRY=True&I=222&J=127"
./build-QGIS-Desktop-Debug/output/bin/qgis_mapserv.fcgi
```

The output is empty.

Please confirm that you also get an empty map when using EPSG:4326 against the very same project.

I'll prepare another project, based on EPSG:4258 (used all over Europe) to use for these tests.

**Comment by elpaso** [comment]
@pblottiere how do you feel about backporting to 3.10? 
I need `exporter.setTransformGeometries( false );` for https://github.com/qgis/QGIS/pull/32596

**Comment by elpaso** [comment]
@pblottiere nevermind, I've found a workaround to trick the JSON exporter.

**Comment by pblottiere** [comment]
Hi @elpaso,

> how do you feel about backporting to 3.10?
> nevermind, I've found a workaround to trick the JSON exporter.

I had planned to backport this bugfix anyway. I'm probably going to do it early next week.

**Pull request merged**